### PR TITLE
DAC-289 Transaction payment by the Customer

### PR DESCRIPTION
**Pull request by amnambiar**
https://input-output.atlassian.net/browse/DAC-289

### CODE REVIEW

File: react-web/src/components/ConnectWallet/ConnectWallet.tsx [hunk 1]
```diff
@@ -40,9 +46,18 @@ const ConnectWallet = () => {
...
     useEffect(() => {
         if (address) {
-            dispatch(getProfileDetails({"address": address, "wallet": wallet}))            
+            (async () => {
+                try {
+                    const response: any = await dispatch(getProfileDetails({"address": address, "wallet": wallet, "walletName": walletName})).catch(handleError)
+                    setWalletLoading(false)
+                } catch(error) {
+                    setWalletLoading(false)
+                    handleError(error)
```
Comment (bogdan-manole) [inline react-web/src/components/ConnectWallet/ConnectWallet.tsx:55]: where `handleError` is declared?

File: react-web/src/components/CreateCertificate/CreateCertificate.tsx [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    const triggerGetCertificate = async () => {
+        setCertifying(true);
+        setShowError("")
+        try {
+            // const walletAddressRes: any = await fetchData.get('/wallet-address').catch(handleError)
+            // const applicationWallet_receiveAddr = walletAddressRes.data;
+            /** For mock */
+            const applicationWallet_receiveAddr = 'addr_test1qz2rzeqq8n82gajfp35enq3mxhaynx6zhuql2c7yaljr25mfaznfszxu8275k6v7n05w5azzmxahfzdq564xuuyg73pqnqtrrc'
```
Comment (bogdan-manole) [inline react-web/src/components/CreateCertificate/CreateCertificate.tsx:75]: there is already an endpoint on the BE for this. why not using it?

File: react-web/src/components/Header/Header.tsx [hunk 3]
```diff
@@ -15,6 +15,23 @@ const Header = () => {
...
+        try {
+          const enabledWallet = await window.cardano[walletNameCache].enable()
+          dispatch(getProfileDetails({"address": addressCache, "wallet": enabledWallet, "walletName": walletNameCache}))
+        } catch(e) {
+          console.log(e)
+        }
+      })()
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
```
Comment (bogdan-manole) [inline react-web/src/components/Header/Header.tsx:33]:
```suggestion
  }, [dispatch])
```

**Review by bogdan-manole — CHANGES_REQUESTED** [review]
I will pause the review on this because i'm not able to start the app anymore ( I don't know if the problem was the same with some other previous approved PR's )
Practically what happens is:
if you type these startup commands:
```
rm -rf node_modules
npm install
npm start
```

1) on my MacOS `aarch64` computer the "Starting development server" hangs for around 1 minute and crashes with out of memory
2) on my Linux `x64` machine the  "Starting development server" hangs around 3 minutes and crashes with the same error

```
Starting the development server...


<--- Last few GCs --->

[2951849:0x5c950d0]   251313 ms: Scavenge 4062.8 (4139.8) -> 4060.8 (4139.8) MB, 17.4 / 0.0 ms  (average mu = 0.887, current mu = 0.773) allocation failure
[2951849:0x5c950d0]   251406 ms: Scavenge 4063.5 (4139.8) -> 4062.0 (4141.3) MB, 12.8 / 0.0 ms  (average mu = 0.887, current mu = 0.773) allocation failure
[2951849:0x5c950d0]   259859 ms: Mark-sweep 4065.0 (4141.3) -> 4062.5 (4151.1) MB, 8352.6 / 0.0 ms  (average mu = 0.779, current mu = 0.059) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb02ec0 node::Abort() [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
 2: 0xa181fb node::FatalError(char const*, char const*) [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
 3: 0xced88e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
 4: 0xcedc07 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
 5: 0xea5ea5  [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
 6: 0xeb557d v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
 7: 0xeb827e v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
 8: 0xe796aa v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
 9: 0x11f2e86 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
10: 0x15e7879  [/home/horus/.nvm/versions/node/v16.13.0/bin/node]
```

File: react-web/src/components/CreateCertificate/CreateCertificate.tsx (hunk 2 above)
Comment (amnambiar) [reply react-web/src/components/CreateCertificate/CreateCertificate.tsx]: Uncommented above lines, and commented this.

File: react-web/src/components/ConnectWallet/ConnectWallet.tsx (hunk 1 above)
Comment (amnambiar) [reply react-web/src/components/ConnectWallet/ConnectWallet.tsx:56]: Missed, fixed now

File: react-web/src/components/Header/Header.tsx (hunk 3 above)
Comment (amnambiar) [reply react-web/src/components/Header/Header.tsx]: Done

**Comment by bogdan-manole** [comment]
> FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
> It worked for me after those steps. I will look into fixing any memory leaks or how to fix those in react , either with the current ticket in hand or in a newer one

a new ticket won't solve the fact that at least in my case this branch doesn't seem to work anymore ( btw, I tried #47 and it works )

**Comment by amnambiar** [comment]
@bogdan-manole Try this for your global node.js, increasing the memory allocation, and it worked for me. Setting it to 4096 wasn't sufficient. We can embed it to the build scripts, maybe we can do that after the initial deployment trials if at all this error comes up.

`export NODE_OPTIONS=--max_old_space_size=8048`

**Comment by bogdan-manole** [comment]
> @bogdan-manole Try this for your global node.js, increasing the memory allocation, and it worked for me. Setting it to 4096 wasn't sufficient. We can embed it to the build scripts, maybe we can do that after the initial deployment trials if at all this error comes up.
> 
> `export NODE_OPTIONS=--max_old_space_size=8048`



> @bogdan-manole Try this for your global node.js, increasing the memory allocation, and it worked for me. Setting it to 4096 wasn't sufficient. We can embed it to the build scripts, maybe we can do that after the initial deployment trials if at all this error comes up.
> 
> `export NODE_OPTIONS=--max_old_space_size=8048`

I tried as you said, but the process of building it's still crashing when it reaches a total RAM `> 5 GB` 

What worries me more it's why would we need such an increase in memory in one PR. In the `master`, a development node would reach a maximum of 512 MB, and now we need an increase of more than 10 times?! 
Also, in master the server it starts in < 10 seconds, and now it takes around one minute,  just to crash ( eating a lot of RAM)

I suspect we now have a library or some code change that makes that, and we rather know what it is and decide if it pays off for such a huge increase of resources

**Comment by amnambiar** [comment]
Agree, the added packages in this PR are the two essentials [@emurgo/cardano-serialization-lib-asmjs](https://www.npmjs.com/package/@emurgo/cardano-serialization-lib-asmjs) and [buffer](https://www.npmjs.com/package/buffer)

**Comment by amnambiar** [comment]
I've opened an issue https://github.com/Emurgo/cardano-serialization-lib/issues/586 within the repository. I've also added in a sample repository with just the transaction bit in place. Let's wait until the reason behind this build error is tracked.

**Comment by amnambiar** [comment]
I've opened an issue https://github.com/Emurgo/cardano-serialization-lib/issues/586 within the repository. I've also added in a sample repository with just the transaction bit in place. Let's wait until the reason behind this build error is tracked.

**Comment by amnambiar** [comment]
@bogdan-manole ,

A fix for the build error is in place. See [this](https://github.com/input-output-hk/dapps-certification/pull/46/commits/21dad9976380e51ffb259b45d1e95fd84c2cda1a) commit. Please run the below in the order, to ensure this fix (within `react-web/`):
```
$ git checkout DAC-289
$ git fetch origin
$ git pull origin DAC-289
$ rm -rf node_modules
$ npm install
$ npm run start
```

Please see if this works at your end as well.

Cc: @RSoulatIOHK

**Review by bogdan-manole — APPROVED** [review]

**Pull request merged**